### PR TITLE
added validation of signature without stripping port number

### LINKED ISSUE #402: Incoming SMS webhook failing validation (non standard SSL port)
Hi,

I have an incoming SMS webhook using HTTPS but on a non standard port (4433). The incoming POST fails validation unless the changes made in #392 are reverted (i.e. the port number left in for validation). The [security notes](https://www.twilio.com/docs/api/security#notes) don't mention non standard ports. Should only the standard port be removed if using HTTPS?

https://github.com/twilio/twilio-python/blob/119fe50863b2e9df89a56711e8d3410288709593/twilio/request_validator.py#L77

### PR DESCRIPTION
**Pull request by eshanholtz**
Fixes #402
PR for DI-321

### CODE REVIEW

File: twilio/request_validator.py [hunk 1]
```diff
@@ -82,22 +96,29 @@ def validate(self, uri, params, signature):
...
         valid_signature_with_port = False
         valid_body_hash = True  # May not receive body hash, so default succeed
 
-
         query = parse_qs(parsed_uri.query)
         if "bodySHA256" in query and isinstance(params, string_types):
             valid_body_hash = compare(self.compute_hash(params), query["bodySHA256"][0])
-            valid_signature = compare(self.compute_signature(uri, {}), signature)
-            valid_signature_with_port = compare(self.compute_signature(parsed_uri, {}), signature)
+            valid_signature = compare(self.compute_signature(uri_without_port, {}), signature)
```
Comment (childish-sambino) [inline twilio/request_validator.py]: Just set `params = {}` here and then you don't need the same calls in the `else` block.

File: twilio/request_validator.py [hunk 2]
```diff
@@ -82,17 +96,29 @@ def validate(self, uri, params, signature):
...
+
         if parsed_uri.scheme == "https" and parsed_uri.port:
-            uri = remove_port(parsed_uri)
+            uri_without_port = remove_port(parsed_uri)
+        elif parsed_uri.scheme == "https":
+            uri_with_port = add_port(parsed_uri, 443)
+        elif parsed_uri.scheme == "http" and parsed_uri.port:
+            uri_without_port = remove_port(parsed_uri)
+        elif parsed_uri.scheme == "http":
+            uri_with_port = add_port(parsed_uri, 80)
```
Comment (childish-sambino) [inline twilio/request_validator.py]: Should target something like this:
```
uri_with_port = add_port(uri)
uri_without_port = remove_port(uri)
```
I think `remove_port` works fine as-is. `add_port` will need to add the appropriate port if it's missing.

Comment (eshanholtz) [reply twilio/request_validator.py]: Making sure I understand this comment: move the logic of adding/removing ports to the appropriate add/remove functions?

Comment (childish-sambino) [reply twilio/request_validator.py]: Correct. They're independent so it makes sense to isolate the logic. And it makes adding tests for each simple.

File: twilio/request_validator.py [hunk 3]
```diff
@@ -21,19 +21,47 @@ def compare(string1, string2):
...
+    :rtype: str
+    """
+    if uri.port:
+        return uri.geturl()
+    
+    port = None
+    if uri.scheme == "https":
+        port = 443
+    elif uri.scheme == "http":
+        port = 80
```
Comment (childish-sambino) [inline twilio/request_validator.py]: What if it's neither?

Comment (eshanholtz) [reply twilio/request_validator.py]: I was wondering about this... what kind of behavior would we want in that situation? Are non http/https uri's ever valid?

Comment (childish-sambino) [reply twilio/request_validator.py]: I think this is fine for now: `port = 443 if uri.scheme == "https" else 80`.

If we have usecases later for default ports for other schemes they can be added.

File: twilio/request_validator.py [hunk 4]
```diff
@@ -21,19 +21,43 @@ def compare(string1, string2):
...
     new_uri = uri._replace(netloc=new_netloc)
+
+    return new_uri.geturl()
+
+
+def add_port(uri):
+    """Add the port number to a URI
+
+    :param uri: parsed URI that Twilio requested on your server
+    :param port: the port number to be added to the URI
```
Comment (childish-sambino) [inline twilio/request_validator.py]: Port is not passed in.

File: twilio/request_validator.py [hunk 5]
```diff
@@ -82,17 +106,19 @@ def validate(self, uri, params, signature):
...
         query = parse_qs(parsed_uri.query)
         if "bodySHA256" in query and isinstance(params, string_types):
             valid_body_hash = compare(self.compute_hash(params), query["bodySHA256"][0])
-            valid_signature = compare(self.compute_signature(uri, {}), signature)
-        else:
-            valid_signature = compare(self.compute_signature(uri, params), signature)
+            params = {}
+
+        valid_signature = compare(self.compute_signature(uri_without_port, params), signature)
+        valid_signature_with_port = compare(self.compute_signature(uri_with_port, params), signature)
```
[the remark this answers is not in the document]
Comment (eshanholtz) [reply twilio/request_validator.py]: 👍

**Pull request merged**